CounterpointGenerator/CounterpointGenerator/Rules/BattutaRule.cs

```diff
@@ -13,27 +13,37 @@ class BattutaRule : IRules
         /**
          * INPUTS: Possibilities, CurrentNote, PreviousCantus, PreviousCounterpoint
          */
-        public List<int> Apply(RuleInput ruleInput)
+        public List<Note> Apply(RuleInput ruleInput)
         {
-            List<int> output = new List<int>(ruleInput.Possibilities);
+            // Do nothing if no previous notes!
+            if (ruleInput.Position == 0)
+            {
+                return ruleInput.Possibilities;
+            }
+
+            List<Note> output = new List<Note>(ruleInput.Possibilities);
             // Check if previous cantus stepped up
-            if (ruleInput.PreviousCantus < ruleInput.CurrentNote)
+            if (ruleInput.PreviousCantus.Pitch < ruleInput.CurrentNote.Pitch)
             {
                 // Check if previous counterpoint was the higher voice
-                if(ruleInput.PreviousCounterpoint > ruleInput.PreviousCantus)
+                if(ruleInput.PreviousCounterpoint.Pitch > ruleInput.PreviousCantus.Pitch)
                 {
                     // Remove upper octave as a possibility
-                    output.Remove(ruleInput.CurrentNote + 13);
+                    Note removeNote = new Note(ruleInput.CurrentNote.Pitch + 13, ruleInput.CurrentNote.Length);
+                    // Remove all items n such that n's pitch is equal to the removeNote pitch
+                    output.RemoveAll(n => n.Pitch == removeNote.Pitch);
                 }
             }
             // Check if previous cantus stepped down
-            else if (ruleInput.PreviousCantus > ruleInput.CurrentNote)
+            else if (ruleInput.PreviousCantus.Pitch > ruleInput.CurrentNote.Pitch)
             {
                 // Check if previous counterpoint was the lower voice
-                if(ruleInput.PreviousCounterpoint < ruleInput.PreviousCantus)
+                if(ruleInput.PreviousCounterpoint.Pitch < ruleInput.PreviousCantus.Pitch)
```

Comment:
You can collapse these conditionals by using an &

Reply:
Resolved 59bf229